Initial build of service-list web component

bellepx0 opened the pull request:
## Chromatic
<!-- This `{{head.ref}}` is a placeholder for a CI job - it will be updated automatically -->
https://{{head.ref}}--65a6e2ed2314f7b8f98609d8.chromatic.com

---
## Configuring this pull request
- [ ] Link to any related issues in the description so they can be cross-referenced.
- [ ] Add the appropriate version patch label (`major`, `minor`, `patch`, or `ignore-for-release`).
    - See [How to choose a version number](https://github.com/department-of-veterans-affairs/component-library#how-to-choose-a-version-number) for guidance.
    - Use `ignore-for-release` if files haven't been changed in a component library package. (ie. only Storybook files)
- [ ] DST Only: Increment the `/packages/core` version number if this will be the last PR merged before a release.
- [ ] Complete all sections below.
- [ ] Delete this section once complete

## Description
Closes [#295](https://github.com/department-of-veterans-affairs/tmf-auth-exp-design-patterns/issues/295)

## Summary
- Added new service-list web component
- Added Storybook file
- Added e2e test for component

## QA Checklist
- [ ] Component maintains 1:1 parity with design mocks
- [ ] Text is consistent with what's been provided in the mocks
- [ ] Component behaves as expected across breakpoints
- [ ] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [ ] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [ ] Tab order and focus state work as expected
- [ ] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [ ] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [ ] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Screenshots
<img width="935" alt="Screenshot 2025-01-31 at 11 07 24 AM" src="https://github.com/user-attachments/assets/b567bf2e-0f75-4d2d-b4f1-1242567b9158" />

## Acceptance criteria
- [ ] QA checklist has been completed
- [ ] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.scss:
```diff
@@ -0,0 +1,57 @@
+}
+.service-list{
+padding: 1rem 2rem 2rem 1rem;
+border: 1px solid #757575;
```
we can use the css variable here to adhere to the VADS design tokens:
`border: 1px solid var(--vads-color-gray-medium);`

bellepx0 replied:
done

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.scss:
```diff
@@ -0,0 +1,57 @@
+    min-width: 40px !important;
+    max-width: 40px !important;
+    color: #ffffff !important;
+    background-color: #162e51 !important;
```
question: How do we make the background color dynamic? Looks like the designs have several different background colors shown for different icons.

bellepx0 replied:
I added the `icon` prop to the class so we can define the background colors for each icon directly in the SCSS file. There might be an alternative approach like dynamically setting the background color inline using a prop such as  `vads-color-hub-life-insurance`, with `life-insurance` being the passed prop. But I’m keeping it simple for now and we can revisit and adjust the implementation later if needed when integrating this into vets-website

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+    return (
+      <Host>
+        <div class="service-list">
+          <a href="/">
```
we will need this to be a prop so the link href can be dynamic

bellepx0 replied:
done. Changed to `<a href={serviceLink}>`

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+          <a href="/">
+            <div class="header">
+              <slot name="icon">
+                <va-icon class="icon" size={3} icon={icon}></va-icon>
```
since the icon is optional, we will need a way to not render that here if the icon identifier isn't passed in.

bellepx0 replied:
done

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+  shadow: true,
+})
+export class VaServiceList {
+  @Prop() serviceDetails: any;  
```
we should get a type created here for the serviceDetails so that it isn't using `any`

bellepx0 replied:
Created an interface for ServiceDetails and for ServiceAction as well to avoid using `any`

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+    try {
+      this.parsedServiceDetails = typeof newValue === 'string' ? JSON.parse(newValue) : newValue;
+    } catch (error) {
+      console.error('Error parsing serviceDetails:', error);
```
There is a `consoleDevError` that might be worth using here instead of console.error

bellepx0 replied:
done

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+      </li>
+    ));
+
+    const actionNeeded = serviceStatus === 'Eligible';
```
I'm not sure about this check here. Would it be more appropriate to instead check if `parsedAction` has `href` and `text` to determine if `actionNeeded` is truthy?

bellepx0 replied:
yes that would be better. Changed the line to `const actionNeeded = parsedAction?.href && parsedAction?.text`

adamwhitlock1 reviewed (commented):
I'm still working on getting this working locally as I'm having environment / node-gyp issues. I've added some comments though from my first read through of the code.

adamwhitlock1 commented on packages/web-components/src/components.d.ts:
```diff
@@ -1540,6 +1540,19 @@ export namespace Components {
     }
+    /**
+     * @componentName Service List
+     * @maturityCategory use with caution
```
this should be set to ` * @maturityCategory caution` and that _should_ fix it's maturity indicator not showing in storybook

adamwhitlock1 commented on packages/storybook/stories/va-service-list.stories.jsx:
```diff
@@ -0,0 +1,57 @@
+const serviceListDocs = getWebComponentDocs('va-service-list');
+
+export default {
+  title: 'Components/Service list',
```
This needs to change to have a capital L so `'Components/Service List'` and then the maturity status indicator circle icon will show correctly in the storybook sidebar nav.

bellepx0 replied:
I got this to work earlier, but when I changed the component name to "Service List Item", the status indicator no longer shows. The maturity category and level are correctly rendering. Need to look into why this is.

bellepx0 replied:
nevermind. this works again!

bellepx0 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+    const serviceListItems = Object.entries(parsedServiceDetails).map(([key, value]) => (
+      <li>
+        <div>
+          <strong>{key.toUpperCase()}:</strong> {value}
```
done

adamwhitlock1 commented on packages/web-components/src/components/va-service-list/va-service-list.tsx:
```diff
@@ -0,0 +1,101 @@
+                <va-icon class="icon" size={3} icon={icon}></va-icon>
+              </slot>
+              <slot name="service-name">
+                <h3 class="service-name">{serviceName}</h3>
```
Should this heading level be dynamic? My guess is that an H3 may not work in every scenario, and we should be able to configure this via a `serviceNameHeadingLevel` prop or similar naming. Creating it with a default level of 3 would also probably be a good default then.

bellepx0 replied:
done

adamwhitlock1 commented:
Last thing from reviewing this initial work:

I _think_ all the slots should be removed. Having named slots means that every slot's content can be replaced by passing in a slot element as a child to the component, which would leave the door open for any markup to be placed in this component.

bellepx0 commented:
> Last thing from reviewing this initial work:
> 
> I _think_ all the slots should be removed. Having named slots means that every slot's content can be replaced by passing in a slot element as a child to the component, which would leave the door open for any markup to be placed in this component.

Makes sense. I removed all the slots.

jamigibbs commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.tsx:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  private adjustChevronPosition() {
+    setTimeout(() => {
```
I don't think I understand completely the purpose of this method loading in `componentDidLoad`. Do you mind adding a screenshot or a little more info about why it's needed?

bellepx0 replied:
The `adjustChevronPosition()` function adjusts the chevron icon to be top-aligned when the service name wraps onto two lines. By default, if the service name fits on a single line, the chevron remains center-aligned. I used `componentDidLoad()` to ensure the component is fully rendered in the DOM before calling `adjustChevronPosition()`. Otherwise, you get an undefined error.

Service name wraps onto 2 lines, chevron is top-aligned
![Screenshot 2025-02-18 at 12 17 16 PM](https://github.com/user-attachments/assets/09271b44-b0c2-4295-9fff-bbd9c8b4d018)

Service name is 1 line, chevron is center-aligned
![Screenshot 2025-02-18 at 12 18 13 PM](https://github.com/user-attachments/assets/2d4ce5cf-e12a-44d8-8beb-3853d0dc2c32)

micahchiang replied:
The approach here feels a little too brittle to me. I don't necessarily think we need the setTimeout because I suspect throwing this in `componentDidLoad` could accomplish the same thing you're trying to achieve. 

Did you try to handle this just with css?

bellepx0 replied:
@micahchiang You're right, thanks! I got rid of the setTimeout. 

I did initially attempt to handle this with CSS alone, but it couldn’t fully resolve the issue where the chevron needed to adjust when the service name wrapped onto 2 lines. It appears that CSS doesn’t reliably detect text wrapping, so I had to resort to using JavaScript to implement this.

jamigibbs replied:
@bellepx0  Have you tried this CSS only solution yet?

```css
va-icon.chevron-icon {
  flex: 1 1 100%;   <-- make that icon's flex column full width
  margin: auto; <-- vertically center it
  text-align: right; <-- align the icon to the far right
}
```

I added the far right alignment because that seems to be what the[ Figma file](https://www.figma.com/design/ZIGDfSb8D5YLBdJavzDdqi/AE-Design-Patterns---Service-list?node-id=246-3229&p=f&t=c2L4f8WxhEZe2rOP-0) is suggesting but you can remove that if not.

![Screenshot 2025-02-20 at 10 01 28 AM](https://github.com/user-attachments/assets/fd1fc047-0f59-4428-b6d8-c2c8a634af05)


![Screenshot 2025-02-20 at 10 03 44 AM](https://github.com/user-attachments/assets/74eeedaa-675a-4e68-8505-d0d90face7dd)

bellepx0 replied:
@jamigibbs Thank you! I tried implementing your solution. Unfortunately, the styling still doesn’t align the chevron icon to the top when the name wraps to 2 lines. As you can see in your screenshot, the chevron is centered with "disability compensation" instead of being top-aligned with the word "disability."

If JavaScript isn’t the ideal solution for this, please let me know. Otherwise, I can reach out to the design team about this issue to see if we can simplify the approach by having the chevron icon always center-aligned.

jamigibbs replied:
> Unfortunately, the styling still doesn’t align the chevron icon to the top when the name wraps to 2 lines

@bellepx0 oh! Ok now I understand. You explained it earlier and I still didn't understand until now. 😄 Thank you for being patient with my repetitive questions. 

Do you mind confirming that the chevron is supposed to be top aligned when the header wraps? I only see it in the [Figma file](https://www.figma.com/design/ZIGDfSb8D5YLBdJavzDdqi/AE-Design-Patterns---Service-list?node-id=246-3229&p=f&t=e53z2YT674D3auqF-0) centered (it would definitely be nicer to only have to use CSS):

![Screenshot 2025-02-20 at 1 40 44 PM](https://github.com/user-attachments/assets/9d58e133-4266-4c7e-b480-86642a81eb70)

bellepx0 replied:
@jamigibbs No worries at all! I think you are looking at an outdated Figma file. This [file](https://www.figma.com/design/ZIGDfSb8D5YLBdJavzDdqi/AE-Design-Patterns---Service-list?node-id=1168-8090&t=2y7b1F1VWFx8WqZ7-1) is the updated version showing the correct chevron alignment. If you scroll towards the right, you should see a heading that says "Final Components (Feb 18, 2025)" (screenshot below)

![Screenshot 2025-02-21 at 8 33 42 AM](https://github.com/user-attachments/assets/f13881ed-20f1-4669-8411-41323980ca70)


Also, I believe I may have resolved this issue! I updated the HTML structure of the header row and was able to adjust the chevron using CSS only. I got removed the JS function as well. Screenshot below.

![Screenshot 2025-02-21 at 8 34 56 AM](https://github.com/user-attachments/assets/bebeb41f-7e9f-4f4f-9b83-3c7d0d6edf86)

jamigibbs commented on packages/web-components/src/components/va-service-list-item/test/va-service-list-item.e2e.ts:
```diff
@@ -0,0 +1,213 @@
+      elementHandle,
+    );
+    expect(shadowInnerHTML).not.toContain('<va-link class="hydrated optional-link"></va-link>');
+  });
```
Can you add axe tests as well? You can take a look at other component examples such as [va-accordion](https://github.com/department-of-veterans-affairs/component-library/blob/70f60ac6e358899e851d752897a1c68764a581f0/packages/web-components/src/components/va-accordion/va-accordion-item.e2e.ts#L60-L94).

bellepx0 replied:
done

jamigibbs commented on packages/storybook/stories/wc-helpers.jsx:
```diff
@@ -258,7 +258,7 @@ export function StoryDocs({ storyDefault, componentName, data, children }) {
   const guidanceName = componentDocs?.guidanceName ?? capitalize(_componentName)?.replaceAll('-', ' ');
 
-
+  
```
Very small nit: Can you remove this space so that this file doesn't register a change?

bellepx0 replied:
done

jamigibbs commented on packages/storybook/stories/va-service-list-item.stories.jsx:
```diff
@@ -0,0 +1,162 @@
+  },
+};
+
+export const ServiceListWithMultipleServiceListItems = ({}) => {
```
For some reason the [code example for this story](https://create-va-service-list-component--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/components-va-service-list-item--docs#service-list-with-multiple-service-list-items) says `<bound Template ...`. Is it possible to have it display as `va-service-list-item` or the React binding `VaServiceListItem`? It might be confusing how to use that example otherwise.

![Screenshot 2025-02-17 at 9 09 32 AM](https://github.com/user-attachments/assets/029a1211-099c-4199-8ed3-5c998dc26d10)

bellepx0 replied:
Fixed. I chose to manually fix the Storybook code preview for this because Storybook does not correctly interpret dynamically rendered web components inside arrays or mapped JSX. Let me know if this is alright to do.

jamigibbs commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.tsx:
```diff
@@ -0,0 +1,183 @@
+  @Prop() serviceLink: string;
+
+  /** The status of the service */
+  @Prop() serviceStatus: string;
```
Several of the properties are not displaying their descriptions on the [Storybook page](https://create-va-service-list-component--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/components-va-service-list-item--docs#service-list-with-multiple-service-list-items). Is it possible to have those displaying? 

![Screenshot 2025-02-17 at 9 12 46 AM](https://github.com/user-attachments/assets/1e8fbb68-8eda-477e-928d-1117eb5a90f4)

bellepx0 replied:
Done. The prop descriptions do appear in the "Properties" section below the Controls. 
![Screenshot 2025-02-18 at 11 45 10 AM](https://github.com/user-attachments/assets/833e815c-4862-440b-98e4-8c6cee9853ec)
After reviewing other components (ex. Select), I noticed that their Controls section also lack descriptions.
![Screenshot 2025-02-18 at 11 45 45 AM](https://github.com/user-attachments/assets/2e797857-1efb-4ace-9568-67969474c839)

I manually added descriptions for each prop in the Controls section. Let me know if this is alright.

jamigibbs commented:
Thank you @bellepx0 for your efforts! I've added the design system team designers in case they need to take a look (possibly also get the Design System Figma files updated?). Also adding Ryan Smith if we need to do an initial accessibility review.

rsmithadhoc commented:
> Thank you @bellepx0 for your efforts! I've added the design system team designers in case they need to take a look (possibly also get the Design System Figma files updated?). Also adding Ryan Smith if we need to do an initial accessibility review.

Yeah, I'll do a [full review](https://design.va.gov/about/accessibility/accessibility-testing-for-design-system-components) of this one. I'll see if I can get to that today or tomorrow.

bellepx0 commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.scss:
```diff
@@ -0,0 +1,163 @@
+@import '../../mixins/uswds-error-border.scss';
+@import '../../mixins/focusable.css';
+@import '../../mixins/headers.css';
+@import '~@department-of-veterans-affairs/css-library/dist/tokens/scss/variables.scss';
```
done

bellepx0 commented:
> Thank you @bellepx0 for your efforts! I've added the design system team designers in case they need to take a look (possibly also get the Design System Figma files updated?). Also adding Ryan Smith if we need to do an initial accessibility review.

Hi @jamigibbs! Thank you for reviewing this. I addressed your comments.

rsmithadhoc reviewed (changes requested):
Hi @bellepx0,

I did my accessibility review, and it looked good for the most part. I have a few small things to report:
- For the heading level prop, I believe we typically make it a `number` value, instead of a string including the `h`. Can you update that so it is consistent with other components?
    - I also was not able to get the Storybook values for heading level and the others in the Properties section to update the example on change. They may not be getting set on the example as expected.
- The heading link "Education" is missing a focus outline. This should have the default yellow outline.
- Yellow focus outline on yellow action link does not have enough contrast.
    - While our current yellow focus isn't perfect in all scenarios, the yellow on yellow is on the extreme end of low contrast, so we should address it.
    - We are working on global focus outline updates, but in the meantime, I think we can override this to have more contrast. `--vads-color-bas` as the focus outline will work for just this link, since this will only be temporary until we update the global focus styles.
- The chevron icon moves when hovering or focusing the heading link, I'm not sure what is causing that. I think we should remove any unnecessary motion.

Thanks!

---

@micahchiang do we need a yellow variation on [Link - Action](https://design.va.gov/storybook/?path=/docs/components-va-link-action--docs) or should this remain a one-off?

bellepx0 commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.scss:
```diff
@@ -0,0 +1,163 @@
+      width: auto;
+    }
+    .chevron-icon {
+      @media (min-width: 767px) {
```
done

bellepx0 commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.scss:
```diff
@@ -0,0 +1,163 @@
+    width: 40px !important;
+    min-width: 40px !important;
+    max-width: 40px !important;
+    color: #ffffff !important;
```
done

bellepx0 commented on packages/web-components/src/components/va-service-list-item/va-service-list-item.scss:
```diff
@@ -0,0 +1,163 @@
+    height: 40px !important;
+    width: 40px !important;
+    min-width: 40px !important;
+    max-width: 40px !important;
```
Fixed. Thanks for catching that. I had leveraged this styling from the vets-website

bellepx0 commented:
> Hi @bellepx0,
> 
> I did my accessibility review, and it looked good for the most part. I have a few small things to report:
> 
> * For the heading level prop, I believe we typically make it a `number` value, instead of a string including the `h`. Can you update that so it is consistent with other components?
>   
>   * I also was not able to get the Storybook values for heading level and the others in the Properties section to update the example on change. They may not be getting set on the example as expected.
> * The heading link "Education" is missing a focus outline. This should have the default yellow outline.
> * Yellow focus outline on yellow action link does not have enough contrast.
>   
>   * While our current yellow focus isn't perfect in all scenarios, the yellow on yellow is on the extreme end of low contrast, so we should address it.
>   * We are working on global focus outline updates, but in the meantime, I think we can override this to have more contrast. `--vads-color-bas` as the focus outline will work for just this link, since this will only be temporary until we update the global focus styles.
> * The chevron icon moves when hovering or focusing the heading link, I'm not sure what is causing that. I think we should remove any unnecessary motion.
> 
> Thanks!
> 
> @micahchiang do we need a yellow variation on [Link - Action](https://design.va.gov/storybook/?path=/docs/components-va-link-action--docs) or should this remain a one-off?

Hi @rsmithadhoc! Thank you for your review. I've addressed your comments as follows: 

1) The heading level prop has been updated to a number value. You can modify the prop values in the Properties section, and they will be reflected in the component.
2) A focus outline has been added to the header row, which includes the icon, service name, and chevron icon, since the entire row is clickable.
3) We are leaving the action link as is for now since it’s FPO. Our team will be working on that specific component next.
4) The chevron icon motion has been removed.

rsmithadhoc reviewed (approved):
Thanks! Approved for accessibility.
